File: quick_sort.py
def partition(arr, last, first):
    i = (last - 1)
    piv = a[first]

    for j in range(last, first):
        if a[j] <= piv:
            i = i + 1
            a[i], a[j] = a[j], a[i]

    a[i + 1], a[first] = a[first], a[i + 1]
    return i + 1


def Quick_Sort(arr, last, first):
    if last < first:
        pi = partition(a, last, first)
        Quick_Sort(a, last, pi - 1)
        Quick_Sort(a, pi + 1, first)


a = []
num = int(input("enter the number of elements in array : "))
print("Enter the elements : ")
for i in range(0, num):
    ele = int(input())
    a.append(ele)


Quick_Sort(a,0,num-1)
print("\nSorted array after QuickSort : ",end=" ")
for i in range(num):
    print(a[i],end=" ")

File: Chapter-5/quick_sort.cpp
#include<bits/stdc++.h>

using namespace std;

void Swap(int *x, int *y){
	int temp;
	temp = *x;
	*x = *y;
	*y = temp;
}

int Partition(int arr[], int last, int first){
	int pivot, index, i;
	index = last;
	pivot = first;

	for(i=last; i < first; i++){
		if(arr[i] < arr[pivot]){
			Swap(&arr[i], &arr[index]);
			index++;
		}
	}
	Swap(&arr[pivot], &arr[index]);
	return index;
}

int random_partition(int arr[], int last, int first){
	int pvt, num, temp;
	num = rand();
	pvt = last + num%(first-last+1);
	Swap(&arr[first], &arr[pvt]);
	return Partition(arr, last, first);
}

int Quick_Sort(int arr[], int last, int first){
	int pindex;
	if(last < first){
		pindex = Partition(arr, last, first);
		Quick_Sort(arr, last, pindex-1);
		Quick_Sort(arr, pindex+1, first);
	}
	return 0;
}

void display(int arr1[],int num){
    cout<<endl<<"sorted Array after QuickSort is: ";
	for (int i=0;i<num;i++){
        cout<<arr1[i]<<" ";
	}
}

int main(){
	int num, i;
	cout<<"enter the number of elements in the array: ";
	cin>>num;

	int arr1[num];
	cout<<"enter elements : ";
	for(i = 0; i < num; i++){
    cin>>arr1[i];
	}

	Quick_Sort(arr1, 0, num-1);
	display(arr1,num);
	return 0;
}
